### ISSUE TITLE
karma fails to serve js test file

### ISSUE THREAD

**Issue by applmak** [bazelbuild/rules_nodejs#1872]
# 🐞 Bug Report!

### Affected Rule

The issue is caused by the rule: karma_web_test

### Is this a regression?

I don't know. I recall having to do something wonky to get karma to work last time I attempted.

### Description

Karma is unable to load files referenced in the 'files' section of the generated configuration file.

## 🔬 Minimal Reproduction

WORKSPACE
```
# Bazel workspace created by @bazel/create 1.6.0

# Declares that this directory is the root of a Bazel workspace.
# See https://docs.bazel.build/versions/master/build-ref.html#workspace
workspace(
    # How this workspace would be referenced with absolute labels from another workspace
    name = "root",
    # Map the @npm bazel workspace to the node_modules directory.
    # This lets Bazel use the same node_modules as other local tooling.
    managed_directories = {"@npm": ["node_modules"]},
)

# Install the nodejs "bootstrap" package
# This provides the basic tools for running and packaging nodejs programs in Bazel
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
http_archive(
    name = "build_bazel_rules_nodejs",
    sha256 = "d14076339deb08e5460c221fae5c5e9605d2ef4848eee1f0c81c9ffdc1ab31c1",
    urls = ["https://github.com/bazelbuild/rules_nodejs/releases/download/1.6.1/rules_nodejs-1.6.1.tar.gz"],
)

# The npm_install rule runs yarn anytime the package.json or package-lock.json file changes.
# It also extracts any Bazel rules distributed in an npm package.
load("@build_bazel_rules_nodejs//:index.bzl", "npm_install")
npm_install(
    # Name this npm so that Bazel Label references look like @npm//package
    name = "npm",
    package_json = "//:package.json",
    package_lock_json = "//:package-lock.json",
)

npm_install(
    # Name this npm so that Bazel Label references look like @npm//package
    name = "latlong_vs_ping_npm",
    package_json = "//latlong_vs_ping:package.json",
    package_lock_json = "//latlong_vs_ping:package-lock.json",
)

# Install any Bazel rules which were extracted earlier by the npm_install rule.
load("@npm//:install_bazel_dependencies.bzl", "install_bazel_dependencies")
install_bazel_dependencies()

# Set up TypeScript toolchain
load("@npm_bazel_typescript//:index.bzl", "ts_setup_workspace")
ts_setup_workspace()

# Fetch transitive Bazel dependencies of npm_bazel_karma
load("@npm_bazel_karma//:package.bzl", "npm_bazel_karma_dependencies")
npm_bazel_karma_dependencies()

# Setup the rules_webtesting toolchain
load("@io_bazel_rules_webtesting//web:repositories.bzl", "web_test_repositories")

web_test_repositories()

load("@io_bazel_rules_webtesting//web/versioned:browsers-0.3.2.bzl", "browser_repositories")

browser_repositories(
    chromium = True,
    firefox = True,
)
```

latlong_vs_ping/BUILD.bazel:
```
load("@npm_bazel_karma//:index.bzl", "karma_web_test_suite")
karma_web_test_suite(
name="test_web_test2",
srcs=["test_web_test.js"],
browsers = [
        "@io_bazel_rules_webtesting//browsers:chromium-local"],
        tags=["native"])
```

test_web_test.js is a trivial test file with a single describe and it block and no content in the it block.

## 🔥 Exception or Error

I've included the long output with high levels of verbosity:
[test.log](https://github.com/bazelbuild/rules_nodejs/files/4584562/test.log)

A snippet:
<pre><code>ESC[36m06 05 2020 02:06:39.846:DEBUG [middleware:source-files]: ESC[39mRequesting /root/latlong_vs_ping/test_web_test.js
ESC[36m06 05 2020 02:06:39.847:DEBUG [middleware:source-files]: ESC[39mFetching /root/latlong_vs_ping/test_web_test.js
ESC[33m06 05 2020 02:06:39.847:WARN [web-server]: ESC[39m404: /root/latlong_vs_ping/test_web_test.js
Chrome Headless 76.0.3809.0 (Mac OS 10.15.4) ERROR: 'There is no timestamp for root/latlong_vs_ping/test_web_test.js!'</code></pre>

## 🌍  Your Environment

**Operating System:**

<pre><code>Mac OS 10.15.4</code></pre>

**Output of `bazel version`:**

<pre><code>Bazelisk version: v1.4.0
Build label: 3.1.0
Build target: bazel-out/darwin-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Wed Apr 22 10:35:52 2020 (1587551752)
Build timestamp: 1587551752
Build timestamp as int: 1587551752</code></pre>

**Rules_nodejs version:**

<pre><code>1.6.1</code></pre>

**Anything else relevant?**

I spent a few hours debugging in the karma rules as well as karma itself. The issue seems to be that while the file listed in the generated .conf.js file is named `root/latlong_vs_ping/test_web_test.js`, the require for that gets intercepted by the karma_require_patcher.js script that's written by a rule in rules_nodejs. The interception causes the file to be found in the runfiles manifest, as expected, and the file is loaded. However, once inside of karma, its dictionary of files (the promised value of filesPromise here: https://github.com/karma-runner/karma/blob/a3d1f11542eab55928bbb0cf4f21cfe34b603006/lib/middleware/source_files.js#L32) has a key for the _full path_ to the file, not the relative path from the runfiles directory (in the attached test output this full path is `/private/var/tmp/_bazel_applmak/c2f9aa229c5dcdcafebbb8098192f7f8/sandbox/darwin-sandbox/3/execroot/root/bazel-out/darwin-fastbuild/bin/latlong_vs_ping/test_web_test2_chromium-local.sh.runfiles/root/latlong_vs_ping/test_web_test2_wrapped_test.conf.js`). This means that the findByPath method returns null, and the web-server reports 404.

My question is then: Why is karma's dictionary of files reference these by full path rather than relative-from-the-runfiles? It would make sense to me that the runner would refer to files in that way.

**Comment by applmak** [comment]
I spent more time looking into this.

Another possible explanation is that context.html (the harness in the browser that runs the tests) should be requesting `/base/root/latlong_vs_ping/test_web_test.js`. IIUC, that would expand to the correct path.

I haven't yet figure out how the `conf.files.served` files end up being requested by that harness (requirejs? concatjs_bundle?), but that's my next avenue of investigation.

**Comment by applmak** [comment]
One last update for today. I've figure out how context.html gets filled out, and the mappings listed there are:
```
[
  "  '/base/npm/node_modules/jasmine-core/lib/jasmine-core/jasmine.js': 'fbc7eaf339ea249d8c912b258548d4e7b4dc180c'",
  "  '/base/npm/node_modules/karma-jasmine/lib/boot.js': '760d54bbca4f739f1f8b252c1636d76201cc4e88'",
  "  '/base/npm/node_modules/karma-jasmine/lib/adapter.js': 'f693fb2743611cfdd0e82f7adbb709f510583603'",
  "  '/base/npm/node_modules/requirejs/require.js': '33bc97e78f4905e72b4c1eb2ca0a4662588443e3'",
  "  '/base/npm/node_modules/karma-requirejs/lib/adapter.js': 'a0765fc3eacb39184893303c3bc2e821cb041787'",
  "  '/base/root/latlong_vs_ping/_test_web_test2_wrapped_test.amd_names_shim.js': '20a2514bed7f9d3e158206ae2bf059fa3373bdfa'",
  "  '/base/root/latlong_vs_ping/test_web_test.js': 'ddb7e7c417c723d7ba21e6773cd0fa3c4b6f83ba'",
  "  '/absolute/private/var/tmp/_bazel_applmak/c2f9aa229c5dcdcafebbb8098192f7f8/sandbox/darwin-sandbox/5/execroot/root/_tmp/4542e652aff1500753912d5f59645bf2/tmp-45221fIA8XwhiJyLV.js': 'abafdaf967b82f8c549638f2f4666c9d9d403648'",
  "  '/absolute/concatjs_bundle.js': 'c4c495f2ad710f525f06ea6b53719c73f98175c4'"
]
```

Which all lgtm.

**Comment by applmak** [comment]
I was finally able to spend a bit of time looking into this today, and I've figured more things out:

- The autogenerated requirejs config contains the bit that is trying to load the file with the bad path. The file contains:
```
      // A simplified version of Karma's requirejs.config.tpl.js for use with Karma under Bazel.
      // This does an explicit `require` on each test script in the files, otherwise nothing will be loaded.
      (function(){
        var runtimeFiles = [
          // BEGIN RUNTIME FILES
          
          // END RUNTIME FILES
        ].map(function(file) { return file.replace(/\.js$/, ''); });
        var allFiles = [
            // BEGIN USER FILES
            'root/latlong_vs_ping/test_web_test.js',
            // END USER FILES
        ];
        var allTestFiles = [];
        allFiles.forEach(function (file) {
          if (/[^a-zA-Z0-9](spec|test)\.js$/i.test(file) && !/\/node_modules\//.test(file)) {
            allTestFiles.push(file.replace(/\.js$/, ''))
          }
        });
        require(runtimeFiles, function() { return require(allTestFiles, window.__karma__.start); });
      })();
```

and it's the BEGIN USER FILES block that requests the bad path. Changing this line to something like `/base/root/latlong_vs_ping/test_web_test.js' makes everything work.

**Comment by applmak** [comment]
q: Any reason I shouldn't propose a PR that adjusts the user files to prepend /base/ before requirejs loading?

**Comment by applmak** [comment]
I was too quick. While that hacks makes things work for a single file, the loading fails as soon as that test file tries to import a second file (say, the thing under test).

**Comment by applmak** [comment]
The failure mode is a super weird requirejs failure mode, where it just doesn't make any progress. No tests are run, no error is reported.

**Comment by applmak** [comment]
Okay, finally got it working! Adding the baseUrl to the requirejs config makes it work:

```
        var allTestFiles = [];
        allFiles.forEach(function (file) {
          if (/[^a-zA-Z0-9](spec|test)\\.js$/i.test(file) && !/\\/node_modules\\//.test(file)) {
            allTestFiles.push(file.replace(/\\.js$/, ''))
          }
        });
+        require.config({
+          // Karma serves files under /base, which is the basePath from your config file
+          baseUrl: '/base',
+        });
        require(runtimeFiles, function() { return require(allTestFiles, window.__karma__.start); });
```

Wow that took a long time.

Anyway, any reason that isn't the default? Where is that typically set?

**Comment by mernst-github** [comment]
I went through the exact same rabbit hole in #1867  and IIUC it is just a symptom of your test file not being a proper named module. Once I turned my test files into modules that problem disappeared:

Here's what I pieced together: karma is not supposed to *request* the file in the first place! The contents of the test file are actually injected into the module system by another generated file that does something like `loadFile("path/to/your_test.js", "contents of your test.js);`, the `require` then call only initializes them. For this to work, the file `your_test.js` needs to actually a named module `your_test` and a plain "describe" is not sufficient for that.

https://requirejs.org/docs/api.html#modulename is the thing that needs to happen, either explicitly, or implicitly by, say, the TS compiler. Note how all typescript examples in this repository use `export {}` to make the test file a module, and I believe it's exactly for this purpose.
